README.md
TO BUILD
==

### Dependencies

1. [NodeJS/NPM](https://nodejs.org) (Latest Stable Version)
2. [Gulp](http://gulpjs.com) (`sudo npm install -g gulp`)

### Setup

1. `cd` to this directory
2. `npm install`
4. `gulp`
5. `npm start`

GENERAL
==

This is a MEAN stack app without the A(ngular). It is the API for [modwatch](http://github.com/ansballard/modwatch). It is fairly simple, just taking in JSON from the [uploader](http://github.com/ansballard/modwatchuploader), storing it at mongolab, and serving it out via API routes.

The general shape of the API is taking in data through `/loadorder`, which is only hit by the uploader. This route takes username, password, some descriptive fields, and file contents from Skyrim mod files as arrays. This is stored in MongoDB. The data is retrieved by the frontend site, which will pull anything from user profile data (types of files, username, last upload timestamp) to actual file data. All API routes are open, all data is available.

CONTRIBUTING
==

1. Fork this repo
2. Open an issue for the problem/enhancement you want to work on
3. Create a branch that has to do with the issue you want to fix
4. Implement your changes
5. Make a pull request to this repo
6. If there are no merge conflicts, and I've already approved the issue you created, I'll most likely merge your changes in

When making changes, do your best to follow the standards already set in other parts of the repo. Changes should not be noticeable when looking through source code. I would prefer all changes pass `eslint` with the `.eslintrc` in the root directory.

LINKS
==

- [The Live Site](http://www.modwat.ch)
- [The Nexus Mods Page](http://nexusmods.com/skyrim/mods/56640)
- [The Frontend](http://github.com/ansballard/modwatch)
- [The Uploader](http://github.com/ansballard/modwatchuploader)


CHANGELOG.md
# Changelog

## 1.0.2 ()

**Implemented Enhancements**

- Added babel
- Moved built node files to `/dist/`
- Added `server.js` to call node files from root directory
- Added lint warning for unhandled errors in callbacks

## 1.0.1 (9/30/2015)

**Implemented Enhancements**

- Added Changelog
- Added Readme


src/app.js
var express = require("express");
  var bodyParser = require("body-parser");
  var cookieParser = require("cookie-parser");
  var methodOverride = require("method-override");
  var session = require("express-session");
  var cors = require("cors");
  var morgan = require("morgan");
var app = express();

var http = require("http");
var path = require("path");
var jwt = require("jsonwebtoken");
var mongoose = require("mongoose/");

var configDB;

if(process.env.OPENSHIFT_NODEJS_PORT || process.env.OPENSHIFT_NODEJS_IP) {
	configDB = require("../config/db.js").getNewLive(process.env.DBUSERNAME, process.env.DBPASSWORD);
} else {
	configDB = require("../config/db.js").getDev(process.env.DBUSERNAME, process.env.DBPASSWORD);
	//var configDB = require("../config/db.js").getLocal(null, null);
}

mongoose.connect(configDB);

var ipaddress = process.env.OPENSHIFT_NODEJS_IP || "127.0.0.1";
var port = process.env.OPENSHIFT_NODEJS_PORT || 3001;

var corsOptions = {
  origin: true,
  methods: ["GET", "POST"]
};

app.set("port", port);
app.set("ip", ipaddress);
app.set("views", path.join(__dirname, "views"));
app.set("view engine", "html");
app.use(cors());
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));
app.use(morgan(process.env.OPENSHIFT_NODEJS_IP ? undefined : "dev"));
app.use(methodOverride());
app.use(cookieParser());
app.use(session({secret: process.env.DBEXPRESSSECRET, resave: false, saveUninitialized: false}));
app.use(express.static(path.join(__dirname, "public")));
app.use(cors(corsOptions));

var scriptVersion = {
  "0.2": "0.27",
  "0.3": "0.3.1"
};

require("./routes.js")(app, jwt, scriptVersion);

http.createServer(app).listen(app.get("port"), app.get("ip"), function() { "use strict";
  console.log("Express server listening at " + app.get("ip") + ":" + app.get("port"));
});


Gruntfile.js
module.exports = function(grunt) { "use strict";

  grunt.initConfig({
    pkg: grunt.file.readJSON("package.json"),
    uglify: {
      backend: {
        options: {
          banner: "/*! <%= pkg.name %> <%= grunt.template.today(\"dd-mm-yyyy\") %> */\n"
        },
        files: [
          {
            expand: true,
            cwd: "src/",
            src: ["*.js"],
            dest: "./",
            ext: ".min.js"
          }
        ]
      }
    },
    eslint: {
      target: ["src/*.js"]
    },
    watch: {
      backend: {
        files: ["src/*.js"],
        tasks: ["eslint", "uglify:backend"]
      }
    },
    nodemon: {
      dist: {
        script: "app.min.js"
      }
    },
    concurrent: {
      dist: {
        tasks: [
          "nodemon",
          "watch"
        ],
        options: {
          logConcurrentOutput: true
        }
      }
    }
  });

  require("jit-grunt")(grunt);

  grunt.registerTask("default", ["eslint", "uglify:backend"]);
  grunt.registerTask("serve", ["default", "concurrent:dist"]);
};


src/modlist.js
var mongoose = require("mongoose");
var bcrypt = require("bcrypt-nodejs");

var Schema = mongoose.Schema;

var modlistSchema = new Schema({
	list: String, // deprecated
	modlisttxt: String, // deprecated
	skyrimini: String, // deprecated
	skyrimprefsini: String, // deprecated
	username: String,
	password: String,
	plugins: [],
	modlist: [],
	ini: [],
	prefsini: [],
	skse: [],
	enblocal: [],
	tag: String,
	enb: String,
	game: String,
	pic: String,
	badge: String,
	token: String,
	timestamp: Date,
	score: {type: Number, default: 0},
	votedon: [{"username": String, "upvoted": Boolean}]
}, {
	collection: "modlist"
});

modlistSchema.methods.generateHash = function(_password) { "use strict";
	return bcrypt.hashSync(_password, bcrypt.genSaltSync(8), null);
};

modlistSchema.methods.validPassword = function(_password) { "use strict";
	return bcrypt.compareSync(_password, this.password);
};

modlistSchema.methods.shrinkArrays = function shrinkArrays() { "use strict";
	var tempNew = [];
	var save = false;
	var i = 0;

	if(this.plugins && this.plugins.length > 0 && typeof this.plugins[0].name !== "undefined") {
		tempNew = [];
		for(i = 0; i < this.plugins.length; i++) {
			tempNew.push(this.plugins[i].name);
		}
		this.plugins = tempNew;
		save = true;
	}
	if(this.modlist && this.modlist.length > 0 && typeof this.modlist[0].name !== "undefined") {
		tempNew = [];
		for(i = 0; i < this.modlist.length; i++) {
			tempNew.push(this.modlist[i].name);
		}
		this.modlist = tempNew;
		save = true;
	}
	if(this.ini && this.ini.length > 0 && typeof this.ini[0].name !== "undefined") {
		tempNew = [];
		for(i = 0; i < this.ini.length; i++) {
			tempNew.push(this.ini[i].name);
		}
		this.ini = tempNew;
		save = true;
	}
	if(this.prefsini && this.prefsini.length > 0 && typeof this.prefsini[0].name !== "undefined") {
		tempNew = [];
		for(i = 0; i < this.prefsini.length; i++) {
			tempNew.push(this.prefsini[i].name);
		}
		this.prefsini = tempNew;
		save = true;
	}
	if(this.skse && this.skse.length > 0 && typeof this.skse[0].name !== "undefined") {
		tempNew = [];
		for(i = 0; i < this.skse.length; i++) {
			tempNew.push(this.skse[i].name);
		}
		this.skse = tempNew;
		save = true;
	}
	if(this.enblocal && this.enblocal.length > 0 && typeof this.enblocal[0].name !== "undefined") {
		tempNew = [];
		for(i = 0; i < this.enblocal.length; i++) {
			tempNew.push(this.enblocal[i].name);
		}
		this.enblocal = tempNew;
		save = true;
	}
	if(save) {
		this.save();
	}
};

modlistSchema.methods.updateFile = function updateFile(userFile, type) {
	"use strict";
	var tempOld = [];
	var tempNew = [];
	tempOld = userFile.split("\",\"");
	tempOld[0] = tempOld[0].substring(2);
	tempOld[tempOld.length - 1] = tempOld[tempOld.length - 1].substring(0, tempOld[tempOld.length - 1].length - 2);
	for(var i = 0; i < tempOld.length; i++) {
		tempNew[i] = tempOld[i];
	}
	return tempNew;
};

/**
 *  Scoring Logic
 */

modlistSchema.methods.votedOnUser = function upvotedUser(username) {
	"use strict";
	var info = {
		index: -1,
		upvoted: false
	};
	for(var i = 0; i < this.votedon.length; i++) {
		if(this.votedon[i].username === username) {
			info.index = i;
			info.upvoted = this.votedon[i].upvoted;
			break;
		}
	}
	return info;
};

module.exports = mongoose.model("Modlist", modlistSchema);


src/passport.js
var LocalStrategy = require('passport-local').Strategy;

module.exports = function(passport) {

	passport.serializeUser(function(user, done) {
	  done(null, user.id);
	});

	passport.deserializeUser(function(id, done) {
	  Modlist.findById(id, function(err, user) {
	    done(err, user);
	  });
	});

	// Register
	passport.use('register', new LocalStrategy({
		usernameField : 'username',
		passwordField : 'password',
		passReqToCallback : true
	},
	function(req, username, password, done) {
		process.nextTick(function() {
			Modlist.findOne({'username' : username }, function(err, user) {
				if(err) {
					return done(err);
				}
				if(user) {
					return done(null, false, console.log("username already taken"));
				}
				else {
					var user = new Modlist(); // a change

					user.username = username;
					user.password = user.generateHash(password);

					user.save(function(err) {
						if(err) {
							throw err;
						}
						return done(null, user);
					});
				}
			});
		});
	}));

	// Login
	passport.use('login', new LocalStrategy({
		usernameField : 'username',
		passwordField : 'password',
		passReqToCallback : true
	},
	function(req, username, password, done) {
		Modlist.findOne({'username' : username }, function(err, user) {
			if(err) {
				return done(err);
			}
			else if(!user) {
				return done(null, false, console.log("username doesn't exist"));
			}
			else if(!user.validPassword(password)) {
				return done(null, false, console.log("Wrong Password"));

			}

			else {
				return done(null, user);
			}
		});
	}));
};

var Modlist = require('./modlist.min.js');


src/routes.js
var Modlist = require("./modlist.js");

var supportedFiletypes = [
	"plugins",
	"modlist",
	"ini",
	"prefsini",
	"skse",
	"enblocal"
];

var validFiletype = function validFileType(filetype) { "use strict";
	return supportedFiletypes.indexOf(filetype) >= 0;
};

var tokenEnsureAuthorized = function tokenEnsureAuthorized(req, res, next) { "use strict";

	var bearerToken;
	var bearerHeader = req.headers.authorization;
	if (typeof bearerHeader !== "undefined") {
		var bearer = bearerHeader.split(" ");
		bearerToken = bearer[1];
		req.token = bearerToken;
		next();
	} else {
		res.sendStatus(403);
	}
};

module.exports = function(app, jwt, scriptVersion) { "use strict";

  /**
   *  Will need routes for option on each when token auth comes in
   */

	app.get("/api/users/count", function(req, res) {
		Modlist.find({}, {_id: 1}, function(err, _modlists) {
			if(_modlists) {
				res.set("Content-Type", "text/plain");
				res.send("" + _modlists.length);
			} else {
				res.writeHead(404);
				res.end();
			}
		});
	});
	app.get("/api/users/list", function(req, res) {
		Modlist.find({}, {username: 1, timestamp: 1, score: 1}, function(err, _mods) {
			var mods = [];
			for(var i = _mods.length - 1, j = 0; i >= 0; i--, j++) {
				mods[j] = {"username": _mods[i].username, "score": _mods[i].score, "timestamp": _mods[i].timestamp};
			}
			res.set("Content-Type", "application/json");
			res.send(mods);
		});
	});
	app.get("/api/script/version", function(req, res) {
		res.set("Content-Type", "text/plain");
		res.end(scriptVersion["0.2"]);
	});
	app.get("/api/script/version/3", function(req, res) {
		res.set("Content-Type", "text/plain");
		res.end(scriptVersion["0.3"]);
	});
	app.get("/api/user/:username/file/:filetype", function(req, res) {
		if(validFiletype(req.params.filetype)) {
			var filetypeJSON = {};
			filetypeJSON[req.params.filetype] = 1;
			Modlist.findOne({username: req.params.username}, filetypeJSON, function(err, _list) {
				if(!_list) {
					res.writeHead(404);
					res.end();
				} else {
					_list.shrinkArrays();
					res.setHeader("Content-Type", "application/json");
					res.end(JSON.stringify(_list[req.params.filetype]));
				}
			});
		} else {
			res.sendStatus(500);
		}
	});
	app.get("/api/user/:username/rawfile/:filetype", function(req, res) {
		if (validFiletype(req.params.filetype)) {
			var filetypeJSON = {};
			filetypeJSON[req.params.filetype] = 1;
			Modlist.findOne({
				username: req.params.username
			}, filetypeJSON, function(err, _list) {
				if (!_list) {
					res.writeHead(404);
					res.end();
				} else {
					_list.shrinkArrays();
					res.setHeader("Content-Type", "text/plain");
					var textList = [];
					for(var i = 0; i < _list[req.params.filetype].length; i++) {
						textList.push(_list[req.params.filetype][i]);
					}
					res.end(textList.join("\n"));
				}
			});
		} else {
			res.sendStatus(500);
		}
	});
	app.get("/api/user/:username/profile", function(req, res) {
		Modlist.findOne({username: req.params.username}, {tag: 1, enb: 1, badge: 1, timestamp: 1, game: 1, score: 1, _id: 0}, function(err, _list) {
			if(!_list) {
				res.writeHead(404);
				res.end();
			} else {
				res.setHeader("Content-Type", "application/json");
				res.end(JSON.stringify(_list));
			}
		});
	});
	app.get("/api/user/:username/files", function(req, res) {
		Modlist.findOne({username: req.params.username}, {plugins: 1, modlist: 1, ini: 1, prefsini: 1, skse: 1, enblocal: 1, _id: 0}, function(err, _list) {
			if(!_list) {
				res.writeHead(404);
				res.end();
			} else {
				res.setHeader("Content-Type", "application/json");
				var arr = [];
				if(_list.plugins.length > 0) {
					arr.push("plugins");
				} if(_list.modlist.length > 0) {
					arr.push("modlist");
				} if(_list.ini.length > 0) {
					arr.push("ini");
				} if(_list.prefsini.length > 0) {
					arr.push("prefsini");
				} if(_list.skse.length > 0) {
					arr.push("skse");
				} if(_list.enblocal.length > 0) {
					arr.push("enblocal");
				}
				res.end(JSON.stringify(arr));
			}
		});
	});
	app.get("/api/search/file/:filetype/:querystring", function(req, res) {
		if(validFiletype(req.params.filetype)) {
			var filetypeJSON = {username: 1};
			filetypeJSON[req.params.filetype] = 1;
			Modlist.find({}, filetypeJSON, function(err, users) {
				var toReturn = [];
				var queryLower = req.params.querystring.toLowerCase();
				var fileLower;
				for(var i = 0; users && i < users.length; i++) {
					users[i].shrinkArrays();
					for(var j = 0; j < users[i][req.params.filetype].length; j++) {
						fileLower = users[i][req.params.filetype][j].toLowerCase();
						if(fileLower.indexOf(queryLower) >= 0) {
							toReturn.push(users[i].username);
							break;
						}
					}
				}
				res.setHeader("Content-Type", "application/json");
				res.end(JSON.stringify({users: toReturn, length: toReturn.length}));
			});
		} else {
			res.sendStatus(500);
		}

  });
  /*app.get("/api/search/timestamp/:from/:to", function(req, res) {
    Modlist.find({}, {username: 1, timestamp: 1}, function(err, users) {
        var toReturn = [];
        for(var i = 0; users && i < users.length; i++) {
          toReturn.push(users[i].timestamp > );
        }
        res.setHeader("Content-Type", "application/json");
        res.end(JSON.stringify({users: toReturn, length: toReturn.length}));
    });
  });*/

  app.post("/auth/checkToken", function(req, res) {
    jwt.verify(req.body.token, process.env.JWTSECRET, function(err, decoded) {
      if(err) {
        res.writeHead(403);
        res.end();
      } else {
        res.setHeader("Content-Type", "application/json");
				res.end(JSON.stringify({"username": decoded.username}));
      }
    });
  });

  app.post("/auth/signin", function(req, res) {
    Modlist.findOne({"username": req.body.username}, function(err, user) {
      if(err) {
        res.writeHead(500);
        res.end();
      } else {
        if(user && user.validPassword(req.body.password)) {
          user.pic = jwt.sign({"username": user.username, "password": user.password}, process.env.JWTSECRET, {expiresInMinutes: 720});
          user.save(function(saveErr, saveUser) {
            if(saveErr) {
              res.writeHead(500);
              res.end();
            } else {
              res.json({token: saveUser.pic});
            }
          });
        } else {
          //console.log(req.body.username);
          res.writeHead(403);
          res.end();
        }
      }
    });
  });

  app.post("/api/newTag/:username", tokenEnsureAuthorized, function(req, res) {
		jwt.verify(req.token, process.env.JWTSECRET, function(err, decoded) {
			if(err) {
				res.writeHead(403);
				res.end();
			} else {
				Modlist.findOne({username: req.params.username}, function(findErr, _list) {
					if(_list) {
						_list.tag = req.body.tag;
						_list.save(function(saveErr) {
							if(saveErr) {
								res.writeHead(500);
								res.end();
							} else {
								res.statusCode = 200;
								res.end();
							}
						});
					} else {
						res.writeHead(404);
						res.end();
					}
				});
			}
		});
	});
	app.post("/api/newENB/:username", tokenEnsureAuthorized, function(req, res) {
		jwt.verify(req.token, process.env.JWTSECRET, function(err, decoded) {
			if(err) {
				res.writeHead(403);
				res.end();
			} else {
				Modlist.findOne({username: req.params.username}, function(findErr, _list) {
					if(_list) {
						_list.enb = req.body.enb;
						_list.save(function(saveErr) {
							if(saveErr) {
								res.writeHead(500);
								res.end();
							} else {
								res.statusCode = 200;
								res.end();
							}
						});
					} else {
						res.writeHead(404);
						res.end();
					}
				});
			}
		});
	});
	app.post("/auth/upvote/:votee", tokenEnsureAuthorized, function(req, res) {
		jwt.verify(req.token, process.env.JWTSECRET, function(jwtVerifyErr, decoded) {
			if(jwtVerifyErr) {
				res.writeHead(403);
				res.write("jwt verification error");
				res.end();
			} else if(decoded.username === req.params.votee) {
				res.writeHead(403);
				res.write("You can't vote for yourself");
				res.end();
			} else if(decoded) {
				Modlist.findOne({"username": decoded.username}, function(voterFindErr, voter) {
					var voterInfo = voter.votedOnUser(req.params.votee);
					if(voter && !voterInfo.upvoted) {
						Modlist.findOne({"username": req.params.votee}, function(findVoteeErr, votee) {
							if(votee) {
								if(voterInfo.index === -1) {
									votee.score += 1;
								} else {
									votee.score += 2;
								}
								votee.save(function saveVotee(voteeSaveErr) {
									if(voteeSaveErr) {
										res.writeHead(500);
										res.write("Error saving votee info");
										res.end();
									} else {
										res.setHeader("Content-Type", "application/json");
										res.end(JSON.stringify({"score": votee.score}));
									}
								});
								if(voterInfo.index !== -1) {
									voter.votedon[voterInfo.index].upvoted = true;
								} else {
									voter.votedon.push({"username": req.params.votee, "upvoted": true});
								}
								voter.save(function saveVoter(saveVoterErr) {
									if(saveVoterErr) {
										console.log(saveVoterErr);
									} else {
										console.log("Saved voter.votedon");
									}
								});
							} else {
								res.writeHead(404);
								res.write("jwt verification error");
								res.end();
							}
						});
					} else {
						res.writeHead(403);
						if(!voter) {
							res.write("Voter not found");
						} else {
							res.write("Voter already voted");
						}
						res.end();
					}
				});
			} else {
				res.writeHead(403);
				res.write("Voter not found");
				res.end();
			}
		});
	});
	app.post("/auth/downvote/:votee", tokenEnsureAuthorized, function(req, res) {
		jwt.verify(req.token, process.env.JWTSECRET, function(jwtVerifyErr, decoded) {
			if(jwtVerifyErr) {
				res.writeHead(403);
				res.write("jwt verification error");
				res.end();
			} else if(decoded.username === req.params.votee) {
				res.writeHead(403);
				res.write("Why would you downvote yourself?");
				res.end();
			} else if(decoded) {
				Modlist.findOne({"username": decoded.username}, function(voterFindErr, voter) {
					var voterInfo = voter.votedOnUser(req.params.votee);
					if(voter && (voterInfo.index === -1 || voterInfo.upvoted)) {
						Modlist.findOne({"username": req.params.votee}, function(findVoteeErr, votee) {
							if(votee) {
								if(voterInfo.index === -1) {
									votee.score -= 1;
								} else {
									votee.score -= 2;
								}
								votee.save(function saveVotee(voteeSaveErr) {
									if(voteeSaveErr) {
										res.writeHead(500);
										res.write("Error saving votee info");
										res.end();
									} else {
										res.setHeader("Content-Type", "application/json");
										res.end(JSON.stringify({"score": votee.score}));
									}
								});
								if(voterInfo.index !== -1) {
									voter.votedon[voterInfo.index].upvoted = false;
								} else {
									voter.votedon.push({"username": req.params.votee, "upvoted": false});
								}
								voter.save(function saveVoter(saveVoterErr) {
									if(saveVoterErr) {
										console.log(saveVoterErr);
									} else {
										console.log("Saved voter.votedon");
									}
								});
							} else {
								res.writeHead(404);
								res.write("jwt verification error");
								res.end();
							}
						});
					} else {
						res.writeHead(403);
						if(!voter) {
							res.write("Voter not found");
						} else {
							res.write("Voter already voted");
						}
						res.end();
					}
				});
			} else {
				res.writeHead(403);
				res.write("Voter not found");
				res.end();
			}
		});
	});
	app.post("/loadorder", function(req, res) {
		Modlist.findOne({"username": req.body.username}, function(err, _modlist) {
			if(_modlist) { // if the username exists in the db
				if(_modlist.validPassword(req.body.password)) {
					if(_modlist.list || _modlist.modlisttxt || _modlist.skyrimini || _modlist.skyrimprefsini) {
						_modlist.list = undefined;
						_modlist.modlisttxt = undefined;
						_modlist.skyrimini = undefined;
						_modlist.skyrimprefsini = undefined;
					}
					_modlist.plugins = req.body.plugins;
					_modlist.modlist = req.body.modlist;
					_modlist.ini = req.body.ini;
					_modlist.prefsini = req.body.prefsini;
					_modlist.skse = req.body.skse;
					_modlist.enblocal = req.body.enblocal;
					_modlist.enb = req.body.enb;
					_modlist.game = req.body.game;
					_modlist.tag = req.body.tag;
					_modlist.timestamp = Date.now();
					_modlist.save(function(saveErr) {
						if(saveErr) {
							res.statusCode = 500;
							res.write("Saving to server failed");
							res.end();
						} else {
							res.statusCode = 200;
							res.write("Access denied, incorrect password");
							res.end();
						}
					});
					//res.setHeader("Content-Type", "application/json");
					//res.end(JSON.stringify({body: req.body, modlist: _modlist}));
				}
				else {
					res.statusCode = 403;
					res.write("Access denied, incorrect password");
					res.end();
				}
			}
			else { // if the username does not exist
				var modlist = new Modlist();
				modlist.plugins = req.body.plugins;
				modlist.modlist = req.body.modlist;
				modlist.ini = req.body.ini;
				modlist.prefsini = req.body.prefsini;
				modlist.skse = req.body.skse;
				modlist.enblocal = req.body.enblocal;
				modlist.enb = req.body.enb;
				modlist.game = req.body.game;
				modlist.tag = req.body.tag;
				modlist.timestamp = Date.now();
				modlist.username = req.body.username;
				modlist.password = modlist.generateHash(req.body.password);
				modlist.save(function(saveErr) {
					if(saveErr) {
						res.statusCode = 500;
						res.end();
					}
					else {
						res.statusCode = 200;
						res.end();
					}
				});
				//res.setHeader("Content-Type", "application/json");
				//res.end(JSON.stringify(modlist));
			}
		});
	});
	app.post("/fullloadorder", function(req, res) {
		Modlist.findOne({"username": req.body.username}, function(err, _modlist) {
			if(_modlist) { // if the username exists in the db
				if(_modlist.validPassword(req.body.password)) {
					_modlist.plugins = _modlist.updateFile(req.body.plugins, "plugins");
					_modlist.modlist = _modlist.updateFile(req.body.modlisttxt, "modlist");
					_modlist.ini = _modlist.updateFile(req.body.skyrimini, "ini");
					_modlist.prefsini = _modlist.updateFile(req.body.skyrimprefsini, "prefsini");
					_modlist.timestamp = Date.now();
					_modlist.save(function(saveErr) {
						if(saveErr) {
							res.statusCode = 500;
							res.end();
						} else {
							res.statusCode = 200;
							res.end();
						}
					});
				}
				else {
					res.statusCode = 403;
					res.write("Access denied, incorrect password");
					res.end();
				}
			}
			else { // if the username does not exist
				var modlist = new Modlist();
				modlist.plugins = modlist.updateFile(req.body.plugins, "plugins");
				modlist.modlist = modlist.updateFile(req.body.modlisttxt, "modlist");
				modlist.ini = modlist.updateFile(req.body.skyrimini, "ini");
				modlist.prefsini = modlist.updateFile(req.body.skyrimprefsini, "prefsini");
				modlist.username = req.body.username;
				modlist.password = modlist.generateHash(req.body.password);
				modlist.timestamp = Date.now();
				modlist.save(function(saveErr) {
					if(saveErr) {
						res.statusCode = 500;
						res.end();
					}
					else {
						res.statusCode = 200;
						res.end();
					}
				});
			}
		});
	});
};
